### ISSUE TITLE
Multi-GPU training with cache=True loads 'n' copies of dataset in RAM

### ISSUE THREAD

**Issue by shubham-lica** [ultralytics/ultralytics#22307]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi Team, 
When training with multiple GPUs using DDP and `cache=True`, each GPU process creates its own copy of the entire cached dataset in RAM, leading to memory exhaustion on large datasets.

### Expected Behavior
With 8 GPUs and a x-GB dataset using `cache=True`, expected memory usage: ~x-GB total (shared cache)

### Actual Behavior
With 8 GPUs and a x-GB dataset using `cache=True`, actual memory usage: ~x*8 GB+ (8 independent caches)
Result: OOM crash on systems

### Setup

I'm currently training a large dataset 
Number of image-label pairs - 368k
Size of the data on RAM 
Ultralytics print on terminal for the size: `train: Caching images (205.4GB RAM): 100% ━━━━━━━━━━━━ 368462/368462 373.5it/s 16:27`
RAM - 1.7 TB
Storage - 5 TB SSD. 

The train parameters are as follows: 

```
MODEL_PATH     = "yolo11x-obb.pt"
DATASET_PATH   = "data_config.yaml"
WANDB_PROJECT  = "yolo-obb"
WANDB_NAME     = "obb"
NUM_EPOCHS     = 150

results = model.train(
    data=DATASET_PATH,
    epochs=NUM_EPOCHS,
    batch=32,
    imgsz=512,
    project=WANDB_PROJECT, 
    name=WANDB_NAME,       
    device=[0, 1, 2, 3, 4, 5, 6, 7],
    workers=15, 
    cache=True,
    verbose=True,
    hsv_h=0.0,
    hsv_s=0.0,
    hsv_v=0.0,
    translate=0.0,
    scale=0.0,
    fliplr=0.0,
    mosaic=0.0,
    erasing=0.0,
    auto_augment=None,
    # fraction=0.01,
    # resume=True,
)
```

### Issue 
What's happening is that when I run this, it starts off with loading 'n' copies of the dataset into RAM. So rather than just loading the train data once on the RAM of 205 GB it loads it 'n' number of times onto the RAM. 

This is evident from the code also as we spawn N independent Python processes (one per GPU), each with its own isolated memory space - This is the part where I need some confirmation from the ultralytics team. Am I understanding it wrong? Because when I scaled down the training to 2 GPUs, it used some ~500 GB of RAM and actually started working! 

I've attached screenshots of explanation offered through Cursor about the issue to pinpoint it. 

## While running with 8 GPUs: 
My entire 1.7 TB of RAM is maxed out and the process gets killed. Added the screenshot of the tail of that log, and here was the tail of the RAM log before it crashed: 

2025-10-07 13:24:16 | RAM: 1701187MB used out of 1703351MB
2025-10-07 13:34:43 | RAM: 1701198MB used out of 1703351MB
2025-10-07 13:44:58 | RAM: 1701208MB used out of 1703351MB
2025-10-07 13:52:07 | RAM: 1701142MB used out of 1703351MB
2025-10-07 13:52:22 | RAM: 1701190MB used out of 1703351MB
2025-10-07 13:52:57 | RAM: 1700720MB used out of 1703351MB
2025-10-07 13:53:19 | RAM: 1700994MB used out of 1703351MB
2025-10-07 13:54:24 | RAM: 1701050MB used out of 1703351MB
2025-10-07 13:55:19 | RAM: 1701165MB used out of 1703351MB
2025-10-07 13:57:44 | RAM: 1700863MB used out of 1703351MB

<img width="1489" height="843" alt="Image" src="https://github.com/user-attachments/assets/370cb269-38f0-4243-848d-10a4acb3699a" />

## While running with 2 GPUs: 

Tail of the training log: 

```
2025-10-07 15:08:12 WARNING ⚠️ cache='ram' may produce non-deterministic training results. Consider cache='disk' as a deterministic alternative if your disk space allows.
val: Caching images (0.4GB RAM): 100% ━━━━━━━━━━━━ 742/742 326.0it/s 2.3s
2025-10-07 15:08:57 Plotting labels to /mnt/lica-data-2/code/ml-hub/layout-analysis/obb/train/yolo-obb/obb48/labels.jpg...
2025-10-07 15:08:57 optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically...
2025-10-07 15:08:57 optimizer: SGD(lr=0.01, momentum=0.9) with parameter groups 173 weight(decay=0.0), 183 weight(decay=0.0005), 182 bias(decay=0.0)
2025-10-07 15:08:57 Image sizes 512 train, 512 val
2025-10-07 15:08:57 Using 30 dataloader workers
2025-10-07 15:08:57 Logging results to /mnt/lica-data-2/code/ml-hub/layout-analysis/obb/train/yolo-obb/obb48
2025-10-07 15:08:57 Starting training for 150 epochs...
2025-10-07 15:08:57
2025-10-07 15:08:57       Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      1/150      10.9G     0.8781      1.529       2.16         87        512: 100% ━━━━━━━━━━━━ 11515/11515 3.7it/s 51:36
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100% ━━━━━━━━━━━━ 47/47 2.7it/s 17.5s
2025-10-07 16:00:51                    all        742       4531      0.412      0.668      0.415      0.343
2025-10-07 16:00:54
2025-10-07 16:00:54       Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
```

Tail of the RAM usage log: 
```
2025-10-07 16:06:41 | RAM: 542864MB used out of 1703351MB
2025-10-07 16:06:46 | RAM: 542814MB used out of 1703351MB
2025-10-07 16:06:51 | RAM: 542893MB used out of 1703351MB
2025-10-07 16:06:56 | RAM: 542836MB used out of 1703351MB
2025-10-07 16:07:02 | RAM: 542913MB used out of 1703351MB
2025-10-07 16:07:07 | RAM: 542864MB used out of 1703351MB
2025-10-07 16:07:12 | RAM: 542879MB used out of 1703351MB
2025-10-07 16:07:17 | RAM: 542882MB used out of 1703351MB
2025-10-07 16:07:22 | RAM: 542928MB used out of 1703351MB
2025-10-07 16:07:27 | RAM: 542948MB used out of 1703351MB
2025-10-07 16:07:32 | RAM: 542887MB used out of 1703351MB
2025-10-07 16:07:37 | RAM: 542865MB used out of 1703351MB
2025-10-07 16:07:42 | RAM: 542933MB used out of 1703351MB
2025-10-07 16:07:47 | RAM: 542922MB used out of 1703351MB
```

Tail of the GPU usage log: 
```
2025-10-07 16:01:38 2025/10/07 16:01:38.811, 0, 70 %, 19 %, 12822 MiB, 81559 MiB
2025-10-07 16:01:38 2025/10/07 16:01:38.811, 1, 65 %, 17 %, 12274 MiB, 81559 MiB
2025-10-07 16:01:38 2025/10/07 16:01:38.811, 2, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:38 2025/10/07 16:01:38.811, 3, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:38 2025/10/07 16:01:38.811, 4, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:38 2025/10/07 16:01:38.811, 5, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:38 2025/10/07 16:01:38.811, 6, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:38 2025/10/07 16:01:38.811, 7, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:43 2025/10/07 16:01:43.812, 0, 86 %, 27 %, 12822 MiB, 81559 MiB
2025-10-07 16:01:43 2025/10/07 16:01:43.813, 1, 63 %, 16 %, 12274 MiB, 81559 MiB
2025-10-07 16:01:43 2025/10/07 16:01:43.813, 2, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:43 2025/10/07 16:01:43.813, 3, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:43 2025/10/07 16:01:43.813, 4, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:43 2025/10/07 16:01:43.813, 5, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:43 2025/10/07 16:01:43.813, 6, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:43 2025/10/07 16:01:43.813, 7, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:48 2025/10/07 16:01:48.814, 0, 57 %, 14 %, 12822 MiB, 81559 MiB
2025-10-07 16:01:48 2025/10/07 16:01:48.814, 1, 64 %, 17 %, 12274 MiB, 81559 MiB
2025-10-07 16:01:48 2025/10/07 16:01:48.814, 2, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:48 2025/10/07 16:01:48.814, 3, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:48 2025/10/07 16:01:48.814, 4, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:48 2025/10/07 16:01:48.814, 5, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:48 2025/10/07 16:01:48.814, 6, 0 %, 0 %, 1 MiB, 81559 MiB
2025-10-07 16:01:48 2025/10/07 16:01:48.814, 7, 0 %, 0 %, 1 MiB, 81559 MiB
```

Please let me know if my understanding is incorrect somewhere. 
I'll dive deeper into it, but any help here would be appreciated. 

### Additional

<img width="809" height="488" alt="Image" src="https://github.com/user-attachments/assets/095fc073-d738-418b-b05b-b4ffe134b6dd" />

<img width="807" height="381" alt="Image" src="https://github.com/user-attachments/assets/2b9e3033-9989-4613-ab1a-d393dd034576" />

<img width="810" height="697" alt="Image" src="https://github.com/user-attachments/assets/265a8a33-df94-4b86-9fef-38cd5edbc5eb" />

<img width="801" height="345" alt="Image" src="https://github.com/user-attachments/assets/30e84004-8202-4d75-856b-5cc1a5b62667" />

**Comment by UltralyticsAssistant** [comment]
👋 Hello @shubham-lica, thank you for your interest in Ultralytics 🚀! This is an automated response to help you get support quickly — an Ultralytics engineer will also assist you here soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. For your multi-GPU + cache=True case, please include:
- Exact command or Python snippet used (including `cache` value, `device`, `workers`, and dataset YAML)  
- `ultralytics` and `torch` versions, CUDA version, OS, and hardware details  
- A small subset dataset or synthetic repro demonstrating RAM scaling with number of GPUs  
- Full console logs from startup through the point RAM spikes, plus `ultralytics.settings` output if customized  
- Whether the behavior differs for `cache='disk'` and for single-process training (`device=0`)  

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a Python>=3.8 environment with PyTorch>=1.8 to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including CUDA/CUDNN, Python and PyTorch preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thanks for the detailed logs and setup notes you’ve already shared 🙌. With an MRE and the additional details above, we can quickly investigate the multi-GPU caching behavior you’re observing.

**Comment by Y-T-G** [comment]
> This is evident from the code also as we spawn N independent Python processes (one per GPU), each with its own isolated memory space - This is the part where I need some confirmation from the ultralytics team. Am I understanding it wrong? Because when I scaled down the training to 2 GPUs, it used some ~500 GB of RAM and actually started working!

DDP launches independent processes, which is why the cache isn't shared. They are different processes and each process has its own memory space. You would need complex explicit memory sharing mechanism to do share the memory at this level
